ext/ddtrace.c

```diff
@@ -1367,6 +1367,131 @@ static void dd_disable_if_incompatible_sapi_detected(void) {
     }
 }
 
+static php_stream_wrapper_ops traced_http_wrapper_wops;
+static php_stream_wrapper_ops traced_https_wrapper_wops;
+static php_stream_wrapper ddtrace_original_http_wrapper;
+static php_stream_wrapper ddtrace_original_https_wrapper;
+static php_stream_wrapper ddtrace_traced_http_wrapper;
+static php_stream_wrapper ddtrace_traced_https_wrapper;
+
+static php_stream *ddtrace_stream_opener(
+    php_stream_wrapper *wrapper,
+    const char *filename,
+    const char *mode,
+    int options,
+    zend_string **opened_path,
+    php_stream_context *context,
+    int flags,
+    const char *persistent_id,
+    uint32_t persistent_id_len,
+    const char *wrapper_data,
+    uint32_t wrapper_data_len
+) {
+    if (!context) {
+        context = php_stream_context_alloc();
+    }
+
+    // Inject distributed tracing headers
+    zval *http_context_zv = zend_hash_str_find(Z_ARRVAL(context->options), "http", sizeof("http") - 1);
+    if (!http_context_zv) {
+        zval tmp;
+        array_init(&tmp);
+        zend_hash_str_update(Z_ARRVAL(context->options), "http", sizeof("http") - 1, &tmp);
+        http_context_zv = zend_hash_str_find(Z_ARRVAL(context->options), "http", sizeof("http") - 1);
+    }
+
+    if (http_context_zv && Z_TYPE_P(http_context_zv) == IS_ARRAY) {
+        HashTable *http_context = Z_ARRVAL_P(http_context_zv);
+        ddtrace_inject_distributed_headers(http_context, HEADER_MODE_CONTEXT);
+    }
+
+    // Open internal span
+    ddtrace_span_data *span = ddtrace_alloc_execute_data_span(-2, NULL);
+    if (span) {
+        ddtrace_set_global_span_properties(span);
+
+        zend_array *meta = ddtrace_property_array(&span->property_meta);
+        zval zv;
+
+        ZVAL_STRING(&zv, "php.stream");
+        zend_hash_str_update(meta, ZEND_STRL("component"), &zv);
+
+        ZVAL_STRING(&zv, "client");
+        zend_hash_str_update(meta, ZEND_STRL("span.kind"), &zv);
+
+        ZVAL_STRING(&zv, filename);
+        zend_hash_str_update(meta, ZEND_STRL("http.url"), &zv);
+
+        zval *method_zv = zend_hash_str_find(Z_ARRVAL(context->options), "method", sizeof("method") - 1);
+        if (method_zv && Z_TYPE_P(method_zv) == IS_STRING) {
+            zend_hash_str_update(meta, ZEND_STRL("http.method"), method_zv);
+        }
+
+        const char *host_start = strstr(filename, "://");
+        if (host_start) {
+            host_start += 3; // skip "://"
+            const char *host_end = strchr(host_start, '/');
+            size_t host_len = host_end ? (size_t)(host_end - host_start) : strlen(host_start);
+            char *host = estrndup(host_start, host_len);
+            ZVAL_STRING(&zv, host);
+            zend_hash_str_update(meta, ZEND_STRL("network.destination.name"), &zv);
+            efree(host);
+        }
+    }
+
+    php_stream_wrapper *original_wrapper =
+        strncmp(filename, "https://", 8) == 0 ? &ddtrace_original_https_wrapper : &ddtrace_original_http_wrapper;
```

Comment:
It would be cleaner to do two functions `ddtrace_stream_opener_http()` and `ddtrace_stream_opener_https()` which forward the original_wrapper to a common function. I'm not sure either whether that comparison universally holds.